### PR TITLE
Output of GetHistoricRatesAsync not consistent

### PR DESCRIPTION
**Pull request by chrisw000**
See #43 

Serialize instead to IList<Candle>

### CODE REVIEW

**Comment by dougdellolio** [comment]
awesome! i will take a closer look at this and test it out before merging. thanks 😄

**Comment by chrisw000** [comment]
I've got a method that auto handles the batching of historic requests if you want that checked in as well?

File: GDAXClient/Services/Products/Models/CandleGranularity.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿namespace GDAXClient.Services.Products.Models
+{
+    /// <summary>
+    /// The granularity field must be one of the following values: {60, 300, 900, 3600, 21600, 86400}. 
+    /// These values correspond to timeslices representing:
+    /// one minute, five minutes, fifteen minutes, one hour, six hours, and one day, respectively.
+    /// </summary>
+    public enum CandleGranularity : int
```
Comment (dougdellolio) [inline GDAXClient/Services/Products/Models/CandleGranularity.cs]: Nice

File: GDAXClient/Services/Products/Models/CandleGranularity.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿namespace GDAXClient.Services.Products.Models
+{
+    /// <summary>
```
Comment (dougdellolio) [inline GDAXClient/Services/Products/Models/CandleGranularity.cs]: please remove the summary comments here. I try to keep the code clean of this as this information is available in the docs

File: GDAXClient/Services/Products/Models/Candle.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using Newtonsoft.Json;
+using GDAXClient.Utilities;
+
+namespace GDAXClient.Services.Products.Models
+{
+    //  [ time, low, high, open, close, volume ]
+    [JsonConverter(typeof(CandleConverter))]
```
Comment (dougdellolio) [inline GDAXClient/Services/Products/Models/Candle.cs]: have been meaning to do this but haven’t gotten around to it. thx!

File: GDAXClient/Services/Products/ProductsService.cs [hunk 4]
```diff
@@ -84,7 +84,35 @@ public class ProductsService : AbstractService
             return httpResponseMessage;
         }
 
-        public async Task<IEnumerable<object[]>> GetHistoricRatesAsync(ProductType productPair, DateTime start, DateTime end, int granularity)
+        /// <summary>
+        /// Get the Historic Rates and auto batch request when the date range is too large
+        /// </summary>
```
Comment (dougdellolio) [inline GDAXClient/Services/Products/ProductsService.cs]: please update the readme to reflect these changes

File: GDAXClient/Services/Products/ProductsService.cs [hunk 5]
```diff
@@ -84,7 +84,35 @@ public class ProductsService : AbstractService
...
+            const int maxPeriods = 350; // From GDAX docs
+
+            var rc = new List<Candle>();
+
+            DateTime ? dEnd = end;
+            DateTime dStart;
+
+            var maxBatchPeriod = (int)granularity * maxPeriods;
+
+            // Work backwards, so the results are correctly ordered from latest to oldest
```
Comment (dougdellolio) [inline GDAXClient/Services/Products/ProductsService.cs]: can remove these comments. need to think about this logic a little bit more. but seems right to me.

Comment (chrisw000) [reply GDAXClient/Services/Products/ProductsService.cs]: What isn't immediately obvious is that the DateTime values passed in are not directly used to set the start/end periods of the candle data coming back from GDAX

EG if you asked for 5 min data... with a start time of say 10:12 - the data coming back is for the GDAX 5 min candles at 10:15, 10:20, 10:25 etc. (IE not 10:12, 10:17, 10:22 etc.)

So we do need to use the datetime of the last candle of the batch coming back, as the End Time of the next batch... and then set the next batch start time backwards in time from that discovered date.

**Comment by chrisw000** [comment]
I've checked those changes in.


BTW: This the serialize settings I put into the Historic method really needs to go onto all the requests - 
: new JsonSerializerSettings {FloatParseHandling = FloatParseHandling.Decimal});

currently you're losing precision on decimal data that runs to many sig. figs. (albeit an absolutely minuscule amount)

Suggestion would be to leave as is for now...  

#70 moves towards allowing for a single central GET / POST / DELETE etc. method  that does error handling/logging/serialize settings etc.

**Pull request merged**